File: backend/services/feed_service.py
"""Feed service — mixed algorithm reverse-chronological with boosts."""
from __future__ import annotations
import math
from datetime import datetime, timezone, timedelta
from typing import Any
from bson import ObjectId
from fastapi import HTTPException

from database import get_db
from services.listing_service import _serialize
from services import follow_service


TYPE_FILTER = {
    "all": None,
    "products": ["new_product", "old_product"],
    "new_products": ["new_product"],
    "old_products": ["old_product"],
    "services": ["service"],
    "reels": None,  # reels handled separately (reel != null)
}


def _score(listing: dict, dist_km: float | None, following_set: set, now: datetime) -> float:
    s = 0.0
    # Freshness < 24h
    created = listing.get("created_at")
    if isinstance(created, str):
        try:
            c = datetime.fromisoformat(created.replace("Z", "+00:00"))
            if (now - c) < timedelta(hours=24):
                s += 20
        except Exception:  # noqa: BLE001
            pass
    # Proximity (linear 30→0)
    if dist_km is not None:
        s += max(0.0, 30.0 - dist_km)
    # Vendor follow boost
    if listing.get("vendor_id") and str(listing["vendor_id"]) in following_set:
        s += 15
    # Has reel
    if listing.get("reel"):
        s += 10
    # Has offer
    if listing.get("offer_price") is not None:
        s += 5
    # Phase 4b: boost
    boost_exp = listing.get("boost_expires_at")
    if boost_exp:
        try:
            bexp = datetime.fromisoformat(str(boost_exp).replace("Z", "+00:00"))
            if bexp > now:
                s += 25
        except Exception:  # noqa: BLE001
            pass
    return s


async def build_feed(
    type_: str = "all",
    lat: float | None = None,
    lng: float | None = None,
    radius_km: float | None = 10.0,
    cursor: str | None = None,
    limit: int = 20,
    user_id: str | None = None,
    reels_only: bool = False,
) -> dict:
    db = get_db()
    q: dict[str, Any] = {"is_deleted": {"$ne": True}, "status": "active", "is_takendown": {"$ne": True},
                          "title": {"$not": {"$regex": "^TEST_"}}}
    types = TYPE_FILTER.get(type_)
    if types:
        q["type"] = {"$in": types}
    if reels_only or type_ == "reels":
        q["reel"] = {"$ne": None}

    # Pool size — we score client-side after fetching a wider pool
    pool_size = max(limit * 5, 40)

    docs: list[dict]
    if lat is not None and lng is not None and radius_km:
        try:
            # $geoNear via aggregation for distances
            pipeline = [
                {"$geoNear": {
                    "near": {"type": "Point", "coordinates": [float(lng), float(lat)]},
                    "distanceField": "distance_meters",
                    "maxDistance": float(radius_km) * 1000.0,
                    "query": q,
                    "spherical": True,
                }},
                {"$sort": {"_id": -1}},
                {"$limit": pool_size},
            ]
            docs = await db.listings.aggregate(pipeline).to_list(pool_size)
        except Exception:  # noqa: BLE001
            # Fall back to non-geo if index missing / bad data
            docs = await db.listings.find(q).sort([("_id", -1)]).limit(pool_size).to_list(pool_size)
    else:
        docs = await db.listings.find(q).sort([("_id", -1)]).limit(pool_size).to_list(pool_size)

    # Following set for the current user (empty for anon)
    following_set: set = set()
    if user_id:
        following_set = set(await follow_service.following_ids(user_id))

    now = datetime.now(timezone.utc)
    scored = []
    for d in docs:
        dist = d.get("distance_meters")
        dist_km = (dist / 1000.0) if dist is not None else None
        scored.append((_score(d, dist_km, following_set, now), d))

    # Sort by score desc, then _id desc as tiebreak (recency)
    scored.sort(key=lambda x: (x[0], str(x[1]["_id"])), reverse=True)

    # Cursor-based pagination: skip until we're past the cursor _id
    if cursor:
        found = False
        filtered = []
        for _, d in scored:
            if str(d["_id"]) == cursor:
                found = True
                continue
            if found:
                filtered.append(d)
        # If cursor not found in pool, fall through
        chosen = filtered[:limit] if found else [d for _, d in scored][:limit]
    else:
        chosen = [d for _, d in scored][:limit]

    # Enrich with vendor mini and interaction state
    result_items = [_serialize(d) for d in chosen]
    vendor_ids = list({r["vendor_id"] for r in result_items if r.get("vendor_id")})
    if vendor_ids:
        vendors = await db.users.find(
            {"_id": {"$in": [ObjectId(v) for v in vendor_ids]}}
        ).to_list(len(vendor_ids))
        vmap = {str(v["_id"]): v for v in vendors}
        for r in result_items:
            v = vmap.get(r.get("vendor_id", ""))
            if v:
                r["vendor"] = {
                    "id": str(v["_id"]),
                    "name": v.get("name"),
                    "profile_pic": v.get("profile_pic"),
                }

    # Interaction state for current user
    if user_id and result_items:
        from services.interaction_service import user_interaction_state
        state = await user_interaction_state(user_id, [r["id"] for r in result_items])
        for r in result_items:
            r["viewer_state"] = state.get(r["id"], {"liked": False, "saved": False})

    return {
        "items": result_items,
        "next_cursor": str(chosen[-1]["_id"]) if len(chosen) == limit else None,
        "has_more": len(chosen) == limit,
    }

File: backend/src/services/feed.service.js
const Listing = require('../models/Listing');
const User = require('../models/User');
const Reel = require('../models/Reel');
const followService = require('./follow.service');
const { serializeListing } = require('./listing.service');
const { notTestFilter } = require('../utils/helpers');

const TYPE_FILTER = {
  all: null,
  products: ['new_product', 'old_product'],
  new_products: ['new_product'],
  old_products: ['old_product'],
  services: ['service'],
  reels: null,
};

const buildFeed = async ({
  type = 'all',
  lat = null,
  lng = null,
  radius_km = 10.0,
  cursor = null,
  limit = 20,
  user_id = null,
  reels_only = false,
  radiusKm = null,
  userId = null,
  reelsOnly = false
} = {}) => {
  const finalUserId = user_id || userId;
  const finalRadiusKm = radiusKm !== null ? radiusKm : radius_km;
  const finalReelsOnly = reels_only || reelsOnly;

  // If type is 'all' and not reelsOnly, we fetch both Reels and Listings
  const fetchReels = finalReelsOnly || type === 'all' || type === 'reels';
  const fetchListings = !finalReelsOnly && type !== 'reels';

  const qListings = { isDeleted: { $ne: true }, status: 'published', ...notTestFilter() };
  const qReels = { isDeleted: { $ne: true }, status: 'published', ...notTestFilter() };

  const poolSize = Math.max(limit * 5, 40);
  let listingDocs = [];
  let reelDocs = [];

  // Query Listings
  if (fetchListings) {
    if (lat !== null && lng !== null && finalRadiusKm) {
      try {
        const pipeline = [
          {
            $geoNear: {
              near: { type: 'Point', coordinates: [parseFloat(lng), parseFloat(lat)] },
              distanceField: 'distance_meters',
              maxDistance: parseFloat(finalRadiusKm) * 1000.0,
              query: qListings,
              spherical: true,
            },
          },
          { $sort: { _id: -1 } },
          { $limit: poolSize },
        ];
        listingDocs = await Listing.aggregate(pipeline);
      } catch (err) {
        listingDocs = await Listing.find(qListings).sort({ _id: -1 }).limit(poolSize).lean();
      }
    } else {
      listingDocs = await Listing.find(qListings).sort({ _id: -1 }).limit(poolSize).lean();
    }
  }

  // Query Reels
  if (fetchReels) {
    if (lat !== null && lng !== null && finalRadiusKm) {
      try {
        const pipeline = [
          {
            $geoNear: {
              near: { type: 'Point', coordinates: [parseFloat(lng), parseFloat(lat)] },
              distanceField: 'distance_meters',
              maxDistance: parseFloat(finalRadiusKm) * 1000.0,
              query: qReels,
              spherical: true,
            },
          },
          { $sort: { _id: -1 } },
          { $limit: poolSize },
        ];
        reelDocs = await Reel.aggregate(pipeline);
      } catch (err) {
        reelDocs = await Reel.find(qReels).sort({ _id: -1 }).limit(poolSize).lean();
      }
    } else {
      reelDocs = await Reel.find(qReels).sort({ _id: -1 }).limit(poolSize).lean();
    }
  }

  // Combine and label them
  const combined = [];
  for (const doc of listingDocs) {
    combined.push({
      postType: 'listing',
      d: doc,
      createdAt: doc.createdAt || doc.created_at || new Date(0),
    });
  }
  for (const doc of reelDocs) {
    combined.push({
      postType: 'reel',
      d: doc,
      createdAt: doc.createdAt || doc.created_at || new Date(0),
    });
  }

  // Following set
  const followingSet = new Set();
  if (finalUserId) {
    const ids = await followService.followingIds(finalUserId);
    for (const id of ids) {
      followingSet.add(id.toString());
    }
  }

  const now = new Date();
  const scored = [];
  for (const item of combined) {
    const d = item.d;
    const dist = d.distance_meters;
    const distKm = dist !== undefined && dist !== null ? dist / 1000.0 : null;
    
    // Scoring logic
    let score = 0.0;
    const created = item.createdAt;
    if (created && (now - new Date(created) < 24 * 60 * 60 * 1000)) {
      score += 20;
    }
    if (distKm !== null) {
      score += Math.max(0.0, 30.0 - distKm);
    }
    const vendorId = (item.postType === 'reel' ? d.creator : d.vendor)?.toString();
    if (vendorId && followingSet.has(vendorId)) {
      score += 15;
    }
    if (item.postType === 'reel') {
      score += 10;
    }
    if (d.isBoosted || d.boostExpiresAt) {
      score += 25;
    }
    scored.push({ score, item });
  }

  // Sort by score desc, then createdAt desc
  scored.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;
    return new Date(b.item.createdAt) - new Date(a.item.createdAt);
  });

  // Pagination
  let chosen = [];
  if (cursor) {
    let found = false;
    const filtered = [];
    for (const s of scored) {
      const itemId = String(s.item.d._id);
      if (itemId === cursor) {
        found = true;
        continue;
      }
      if (found) {
        filtered.push(s.item);
      }
    }
    chosen = found ? filtered.slice(0, limit) : scored.map(s => s.item).slice(0, limit);
  } else {
    chosen = scored.map(s => s.item).slice(0, limit);
  }

  // Populate vendor/creator details
  const resultItems = [];
  for (const item of chosen) {
    const d = item.d;
    const serialized = item.postType === 'listing' ? serializeListing(d) : { ...d, id: d._id.toString() };
    serialized.postType = item.postType;
    if (d.distance_meters !== undefined) {
      serialized.distance_meters = d.distance_meters;
    }
    resultItems.push(serialized);
  }

  // Populate vendor info
  const listingVendorIds = resultItems.filter(r => r.postType === 'listing' && r.vendor).map(r => r.vendor.toString());
  const reelCreatorIds = resultItems.filter(r => r.postType === 'reel' && r.creator).map(r => r.creator.toString());
  const allUserIds = Array.from(new Set([...listingVendorIds, ...reelCreatorIds]));

  if (allUserIds.length > 0) {
    const users = await User.find({ _id: { $in: allUserIds } })
      .select('name profile_pic avatarUrl')
      .lean();
    const umap = {};
    for (const u of users) {
      umap[u._id.toString()] = u;
    }
    for (const r of resultItems) {
      const uId = r.postType === 'listing' ? r.vendor?.toString() : r.creator?.toString();
      const u = umap[uId];
      if (u) {
        const userObj = {
          id: u._id.toString(),
          _id: u._id.toString(),
          name: u.name,
          profile_pic: u.profile_pic || u.avatarUrl,
          avatarUrl: u.avatarUrl || u.profile_pic,
        };
        if (r.postType === 'listing') {
          r.vendor = userObj;
        } else {
          r.creator = userObj;
        }
      }
    }
  }

  // Populate viewer interactions
  if (finalUserId && resultItems.length > 0) {
    const interactionService = require('./interaction.service');
    const state = await interactionService.userInteractionState(finalUserId, resultItems.map(r => r.id));
    for (const r of resultItems) {
      r.viewer_state = state[r.id] || { liked: false, saved: false };
    }
  }

  return {
    items: resultItems,
    next_cursor: chosen.length === limit ? String(chosen[chosen.length - 1].d._id) : null,
    has_more: chosen.length === limit,
  };
};

module.exports = { buildFeed };